File: android/app/src/main/java/widget/ListProvider.java
package widget;

import android.annotation.TargetApi;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import fhiquo.fhiquo.R;


/*
 *  If you are familiar with Adapter of ListView, this is the same as adapter
 *  with few changes
 */
@TargetApi(Build.VERSION_CODES.HONEYCOMB)
public class ListProvider implements RemoteViewsService.RemoteViewsFactory {
    private Context context;
    private int appWidgetId;
    private String quoteText;
    private String authorText;
    private Integer quoteId;

    public ListProvider(Context context, Intent intent) {
        this.context = context;
        appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        quoteText = intent.getStringExtra(Codes.INTENT_EXTRA_QUOTE_TEXT);
        authorText = intent.getStringExtra(Codes.INTENT_EXTRA_AUTHOR_TEXT);
        quoteId = intent.getIntExtra(Codes.INTENT_EXTRA_QUOTE_ID_WIDGET, -1);
    }

    /*
     * Similar to getView of Adapter where instead of View we return RemoteViews
     */
    @Override
    public RemoteViews getViewAt(int position) {
        final RemoteViews bodyView = new RemoteViews(context.getPackageName(), R.layout.widget_quote_row);
        bodyView.setTextViewText(R.id.BodyText, quoteText);
        bodyView.setTextViewText(R.id.AuthorText, authorText);
        bodyView.setViewVisibility(R.id.AuthorText, IsEmptyOrNull(authorText) ? View.GONE : View.VISIBLE);

        // Next, set a fill-intent, which will be used to fill in the pending intent template
        // that is set on the collection view in StackWidgetProvider.
        Bundle extras = new Bundle();
        extras.putInt(Codes.INTENT_EXTRA_QUOTE_ID_WIDGET, quoteId);
        Intent fillInIntent = new Intent();
        fillInIntent.putExtras(extras);
        // Make it possible to distinguish the individual on-click
        // action of a given item
        bodyView.setOnClickFillInIntent(R.id.widget_quote_frame, fillInIntent);

        return bodyView;
    }

    @Override
    public void onCreate() {

    }

    @Override
    public void onDataSetChanged() {

    }

    @Override
    public void onDestroy() {

    }

    @Override
    public int getCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    public static boolean IsEmptyOrNull(String string) {
        if (string == null)
            return true;

        if (string.length() == 0)
            return true;

        return false;
    }

}

File: android/app/src/main/kotlin/fhiquo/poc/ListProvider.kt
package phiquo.app.widget

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import fhiquo.poc.Codes
import fhiquo.poc.R
import fhiquo.poc.WidgetDatabase
import fhiquo.poc.WidgetQuote


/*
 *  If you are familiar with Adapter of ListView, this is the same as adapter
 *  with few changes
 */
class ListProvider(private val context: Context, intent: Intent) : RemoteViewsService.RemoteViewsFactory {
    private val appWidgetId: Int = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
    private val quoteText: String = intent.getStringExtra(Codes.INTENT_EXTRA_QUOTE_TEXT)
    private val authorText: String = intent.getStringExtra(Codes.INTENT_EXTRA_AUTHOR_TEXT)
    private val updatedText: String = intent.getStringExtra(Codes.INTENT_EXTRA_UPDATED_TEXT)
    private val quoteId: Int = intent.getIntExtra(Codes.INTENT_EXTRA_QUOTE_ID_WIDGET, -1)

    /*
     * Similar to getView of Adapter where instead of View we return RemoteViews
     */
    override fun getViewAt(position: Int): RemoteViews {
        val bodyView = RemoteViews(context.packageName, R.layout.widget_quote_row)

        val db = WidgetDatabase(context);
        val quote = db.getRandomQuote()

        bodyView.setTextViewText(R.id.BodyText, getBody(quote))
        bodyView.setTextViewText(R.id.AuthorText, getAuthor(quote))
        bodyView.setViewVisibility(R.id.AuthorText, if (getAuthor(quote).isNullOrEmpty()) View.GONE else View.VISIBLE)
        bodyView.setTextViewText(R.id.UpdatedText, updatedText)

        return bodyView
    }

    private fun getBody(quote: WidgetQuote): String {
        return if (quote.body != "ERROR") quote.body else quoteText
    }

    private fun getAuthor(quote: WidgetQuote): String {
        return if (quote.author != "ERROR") quote.author else authorText
    }

    override fun onCreate() {}
    override fun getLoadingView(): RemoteViews? = null
    override fun onDataSetChanged() {}
    override fun onDestroy() {}
    override fun getCount(): Int = 1

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getViewTypeCount(): Int {
        return 1
    }

}